### PR TITLE
AB#92107 - Many EndPoints for FrontDoor

### PR DESCRIPTION
**Pull request by PawelBargielski**

### CODE REVIEW

**Review by Carq — APPROVED** [review]

**Review by jjankowski87 — APPROVED** [review]

**Comment by Carq** [comment]
Missing Actions in `HE.Investments.Common.CRM.Model.Actions`

**Review by Carq — COMMENTED** [review]
Missing Actions in HE.Investments.Common.CRM.Model.Actions

**Review by Carq — COMMENTED** [review]
Missing Actions in HE.Investments.Common.CRM.Model.Actions

**Review by Carq — COMMENTED** [review]
Missing Actions in HE.Investments.Common.CRM.Model.Actions

File: src/HE.CRM.Common/DtoMapping/FrontDoorProjectMapper.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+            return frontDoorProjectDto;
+        }
+
+        public static invln_FrontDoorProjectPOC MapDtoToRegularEntity(FrontDoorProjectDto frontDoorProjectDto, Contact contact, string organisationId)
+        {
+            var frontDoorProjectPOC = new invln_FrontDoorProjectPOC()
+            {
+                invln_FrontDoorProjectPOCId = Guid.TryParse(frontDoorProjectDto.ProjectId, out Guid projectid) == true ? projectid : Guid.Empty,
+                //invln_AccountId = new EntityReference(Account.EntityLogicalName, frontDoorProjectDto.OrganisationId),
```
Comment (Antek012) [inline src/HE.CRM.Common/DtoMapping/FrontDoorProjectMapper.cs:82]: comment without TODO ticket

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectRepository.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        {
+        }
+
+        public List<invln_FrontDoorProjectPOC> GetFrontDoorProjectForOrganisationAndContact(Guid organisationId, string externalContactId, string frontDoorProjectId = null, string fieldsToRetrieve = null)
+        {
+            if (frontDoorProjectId == null)
+            {
+                using (DataverseContext ctx = new DataverseContext(service))
+                {
+                    return (from fdp in ctx.invln_FrontDoorProjectPOCSet
```
Comment (Antek012) [inline src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectRepository.cs:26]: Does FieldsToRetrievie feature works for this case?

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            using (var ctx = new OrganizationServiceContext(service))
+            {
+                return ctx.CreateQuery<invln_FrontDoorProjectSitePOC>()
+                    .Where(x => x.invln_FrontDoorProjectId.Id == frontDoorProjectId.Id && x.StateCode.Value == (int)invln_FrontDoorProjectSitePOCState.Active).ToList();
+            }
+        }
+
+        // Maybe it is unnecessary
+
+        //public invln_FrontDoorProjectSitePOC GetSingleSiteRelatedToFrontDoorProject(EntityReference frontDoorProjectId, string frontDoorSiteId)
```
Comment (Antek012) [inline src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs:32]: commented code

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                contactId = ctx.CreateQuery<Contact>().FirstOrDefault(x => x.invln_externalid == externalContactId).Id.ToString();
+            }
+            logger.Trace($"contactid: {contactId}");
+            logger.Trace($"organisationId: {organisationId}");
+
+
+
+            if (Guid.TryParse(frontDoorProjectId, out Guid frontDoorProjecGuid) && Guid.TryParse(organisationId, out Guid organisationGuid))
+            {
+                var fetchXML = @"<fetch version='1.0' output-format='xml-platform' mapping='logical' distinct='false'>
```
Comment (Antek012) [inline src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs:61]: consider using interpolation instead concatenation to improve redability

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	                                <entity name='invln_frontdoorprojectsitepoc'>
+                                            <filter>
+                                              <condition attribute=""statuscode"" operator=""eq"" value=""1"" />
+                                              <condition entityname=""FrontDoorProject"" attribute=""invln_accountid"" operator=""eq"" value=""" + organisationId + @""" />
+                                              <condition entityname=""FrontDoorProject"" attribute=""invln_contactid"" operator=""eq"" value=""" + contactId + @""" />
+                                              <condition attribute=""invln_frontdoorprojectid"" operator=""eq"" value=""" + frontDoorProjectId + @""" />
+                                            </filter>
+		                                    <link-entity name=""invln_frontdoorprojectpoc"" from=""invln_frontdoorprojectpocid"" to=""invln_frontdoorprojectid"" alias=""FrontDoorProject"">
```
Comment (Antek012) [inline src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs:69]: Do we need to define linked entity attributes? Does it have any impact to performance?

File: src/HE.CRM.Plugins/Plugins/CustomApi/GetMultipleFrontDoorProjectsForAccountPlugin.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    [CrmPluginRegistration(
+    "invln_getmultiplefrontdoorprojects",
+    "none",
+    StageEnum.PostOperation,
+    ExecutionModeEnum.Synchronous,
+    "",
+    "HE.CRM.Plugins.Plugins.CustomApi.GetMultipleFrontDoorProjectsForAccountPlugin: invln_getmultiplefrontdoorprojects",
+    1,
+    IsolationModeEnum.Sandbox,
+    Id = "07F3F176-DAC3-4C94-BAA0-D3999CF44FDE")] //done
```
Comment (Antek012) [inline src/HE.CRM.Plugins/Plugins/CustomApi/GetMultipleFrontDoorProjectsForAccountPlugin.cs:20]: comment

File: src/HE.CRM.Plugins/Services/FrontDoorProjectSite/FrontDoorProjectSiteService.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            var frontDoorProjectSiteDto = FrontDoorProjectSiteMapper.MapFrontDoorProjectSiteToDto(frontDoorProjectSite);
+            return frontDoorProjectSiteDto;
+        }
+
+        public List<FrontDoorProjectSiteDto> GetMultipleSiteRelatedToFrontDoorProjectForGivenAccountAndContact(string frontDoorProjectId, string organisationId, string externalContactId)
+        {
+            List<FrontDoorProjectSiteDto> frontDoorProjectSiteDtoList = new List<FrontDoorProjectSiteDto>();
+            var frontDoorProjectSiteList = _frontDoorProjectSiteRepository.GetMultipleSiteRelatedToFrontDoorProjectForGivenAccountAndContact(frontDoorProjectId, organisationId, externalContactId); ;
+
+            if(frontDoorProjectSiteList.Count > 0)
```
Comment (Antek012) [inline src/HE.CRM.Plugins/Services/FrontDoorProjectSite/FrontDoorProjectSiteService.cs:50]: unnecessary if statement

File: src/HE.Common.IntegrationModel/PortalIntegrationModel/FrontDoorProjectDto.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+namespace HE.Common.IntegrationModel.PortalIntegrationModel
+{
+    public class FrontDoorProjectDto
+    {
+        public string ProjectId { get; set; }
+        public string externalId { get; set; }
+        public Guid OrganisationId { get; set; }
+        public string ProjectName { get; set; }
+        public bool? ProjectSupportsHousingDeliveryinEngland { get; set; }
+        public List<int> ActivitiesinThisProject { get; set; } // choises multiselect
```
Comment (Antek012) [inline src/HE.Common.IntegrationModel/PortalIntegrationModel/FrontDoorProjectDto.cs:16]: Are comments needed here?
I think signature (return type and name) should be enough.

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+            }
+            else
+            {
+                if (frontDoorProjecToCreate.invln_ContactId == null)
+                {
+                    frontDoorProjecToCreate.invln_ContactId = requestContact.ToEntityReference();
+                }
+
+                this.TracingService.Trace("Create invln_Loanapplication");
+                //frontDoorProjecToCreate.StatusCode = new OptionSetValue((int)invln_FrontDoorProjectPOC_StatusCode.Active);
```
Comment (Antek012) [inline src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs:135]: comment

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+                }
+                this.TracingService.Trace("GetFrontDoorProjectForOrganisationAndContact");
+                this.TracingService.Trace($"{frontDoorProjectsForAccountAndContact.Count}");
+                foreach (var element in frontDoorProjectsForAccountAndContact)
+                {
+                    List<FrontDoorProjectSiteDto> frontDoorProjectSiteDtoList = new List<FrontDoorProjectSiteDto>();
+                    this.TracingService.Trace($"Front Door Project id {element.Id}");
+                    this.TracingService.Trace("GetFrontDoorProjectSiteRelatedToFrontDoorProject");
+                    var frontDoorProjectSiteList = _frontDoorProjectSiteRepository.GetSiteRelatedToFrontDoorProject(element.ToEntityReference());
+                    if (frontDoorProjectSiteList != null)
```
Comment (Antek012) [inline src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs:75]: is this if needed?

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                            nameof(Contact.Telephone1).ToLower(),
+                        });
+                    }
+                    this.TracingService.Trace("MapRegularEntityToDto");
+                    entityCollection.Add(FrontDoorProjectMapper.MapRegularEntityToDto(element, frontDoorProjectSiteDtoList, externalContactId, frontDoorProjectContact));
+                }
+            }
+
+            this.TracingService.Trace("Serialize");
+            return JsonSerializer.Serialize(entityCollection);
```
Comment (Antek012) [inline src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs:103]: I think Serialize/Deserialize is not a service responsibility.
This should be in handler.

**Review by Antek012 — APPROVED** [review]

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectRepository.cs (hunk 2 above)
Comment (PawelBargielski) [reply src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectRepository.cs:26]: Yes

File: src/HE.CRM.Common/DtoMapping/FrontDoorProjectMapper.cs (hunk 1 above)
Comment (PawelBargielski) [reply src/HE.CRM.Common/DtoMapping/FrontDoorProjectMapper.cs]: Fixed in next Commit

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs (hunk 3 above)
Comment (PawelBargielski) [reply src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs]: Fixed in next Commit

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs (hunk 5 above)
Comment (PawelBargielski) [reply src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs:67]: Yes, the above conditions refer to this table.

File: src/HE.CRM.Plugins/Plugins/CustomApi/GetMultipleFrontDoorProjectsForAccountPlugin.cs (hunk 6 above)
Comment (PawelBargielski) [reply src/HE.CRM.Plugins/Plugins/CustomApi/GetMultipleFrontDoorProjectsForAccountPlugin.cs]: Fixed in next Commit

File: src/HE.Common.IntegrationModel/PortalIntegrationModel/FrontDoorProjectDto.cs (hunk 8 above)
Comment (PawelBargielski) [reply src/HE.Common.IntegrationModel/PortalIntegrationModel/FrontDoorProjectDto.cs]: Fixed in next Commit

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs (hunk 9 above)
Comment (PawelBargielski) [reply src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs]: Fixed in next Commit

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs (hunk 11 above)
Comment (PawelBargielski) [reply src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs]: Fixed in next Commit

File: src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs (hunk 10 above)
Comment (PawelBargielski) [reply src/HE.CRM.Plugins/Services/FrontDoorProject/FrontDoorProjectService.cs]: Fixed in next Commit

File: src/HE.CRM.Plugins/Services/FrontDoorProjectSite/FrontDoorProjectSiteService.cs (hunk 7 above)
Comment (PawelBargielski) [reply src/HE.CRM.Plugins/Services/FrontDoorProjectSite/FrontDoorProjectSiteService.cs]: Fixed in next Commit

File: src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs (hunk 4 above)
Comment (PawelBargielski) [reply src/HE.CRM.Common/Repositories/Implementations/FrontDoorProjectSiteRepository.cs:59]: OK :)

**Pull request merged**